src/lib.rs

```diff
@@ -0,0 +1,489 @@
+    /// on a "best effort" basis -- when the source contains a target dir,
+    /// we use it, otherwise when the other source contains one, we use it,
+    /// otherwise we just use a standard cache folder as specified by XDG.
+    /// We cannot use a temporary directory, because the rustdocs from registry
+    /// are being cached in the target directory.
+    fn get_target_dir(&self, source: &RustdocSource) -> anyhow::Result<PathBuf> {
+        Ok(
+            if let Some(path) = get_target_dir_from_project_root(source)? {
+                path
+            } else if let Some(path) = get_target_dir_from_project_root(&self.current.source)? {
+                path
+            } else if let Some(path) = get_target_dir_from_project_root(&self.baseline.source)? {
+                path
+            } else {
+                get_xdg_cache_dir()
+            },
+        )
+    }
+
+    fn get_rustdoc_generator(
+        &self,
+        config: &mut GlobalConfig,
+        source: &RustdocSource,
+    ) -> anyhow::Result<Box<dyn rustdoc_gen::RustdocGenerator>> {
+        let target_dir = self.get_target_dir(source)?;
+        Ok(match source {
+            RustdocSource::Rustdoc(path) => {
+                Box::new(rustdoc_gen::RustdocFromFile::new(path.to_owned()))
+            }
+            RustdocSource::Root(root) => {
+                Box::new(rustdoc_gen::RustdocFromProjectRoot::new(root, &target_dir)?)
+            }
+            RustdocSource::Revision(root, rev) => {
+                let metadata = manifest_metadata_no_deps(root)?;
+                let source = metadata.workspace_root.as_std_path();
+                Box::new(rustdoc_gen::RustdocFromGitRevision::with_rev(
+                    source,
+                    &target_dir,
+                    rev,
+                    config,
+                )?)
+            }
+            RustdocSource::VersionFromRegistry(version) => {
+                let mut registry = rustdoc_gen::RustdocFromRegistry::new(&target_dir, config)?;
+                if let Some(ver) = version {
+                    let semver = semver::Version::parse(ver)?;
+                    registry.set_version(semver);
+                }
+                Box::new(registry)
+            }
+        })
+    }
+
+    pub fn check_release(&self) -> anyhow::Result<Report> {
+        let mut config = GlobalConfig::new().set_level(self.log_level);
+        let rustdoc_cmd = RustdocCommand::new()
+            .deps(false)
+            .silence(!config.is_verbose());
+
+        let current_loader = self.get_rustdoc_generator(&mut config, &self.current.source)?;
+        let baseline_loader = self.get_rustdoc_generator(&mut config, &self.baseline.source)?;
+
+        let all_outcomes: Vec<anyhow::Result<bool>> = match &self.current.source {
+            RustdocSource::Rustdoc(_)
+            | RustdocSource::Revision(_, _)
+            | RustdocSource::VersionFromRegistry(_) => {
+                let name = "<unknown>";
```

Comment:
In #207 I've made a temporary fix 7ecf5873242bc0defbb1eb739876cdc778614161 so that I can progress further. In the case where this fix is good enough, I can move it here, otherwise we should discuss better alternatives.

Reply:
In https://github.com/obi1kenobi/cargo-semver-checks/commit/7ecf5873242bc0defbb1eb739876cdc778614161 it seems like there's still an `unknown` branch. That looks like it could work if `current` is given by gitrev, but won't work with a registry version nor with a premade rustdoc file.

Can we add an explicit `panic!()` in the cases we know for sure won't work, instead of setting a bogus value and then breaking in a confusing way?